Src/Base/AMReX_BaseFabUtility.H

```diff
@@ -95,6 +95,172 @@ void fill (BaseFab<STRUCT>& aos_fab, F const& f)
     }
 }
 
+//! Transpose 3D array (nx,ny,nz) from row-major (i.e. C order) to
+//! column-major (Fortran order). The input's unit stride direction is z,
+//! whereas the output's unit stride direction is x.
+template <typename T>
+void transposeCtoF (T const* pi, T* po, int nx, int ny, int nz)
+{
+    AMREX_ALWAYS_ASSERT(pi != po);
+
+#if defined(AMREX_USE_CUDA) || defined(AMREX_USE_HIP)
+
+    constexpr int tile_dim = 32;
+    constexpr int block_rows = 8;
+    constexpr int nthreads = tile_dim*block_rows;
+
+    // Each block has tile_dim x block_rows threads. They work on a tile_dim
+    // x tile_dim tile.
+
+    dim3 block{unsigned(tile_dim), unsigned(block_rows), 1};
+    dim3 grid{unsigned((nx+tile_dim-1)/tile_dim),
+              unsigned((nz+tile_dim-1)/tile_dim), unsigned(ny)};
+
+    AMREX_LAUNCH_KERNEL(nthreads, grid, block, 0, Gpu::gpuStream(),
+    [=] AMREX_GPU_DEVICE ()
+    {
+        __shared__ T tile[tile_dim][tile_dim+1]; // +1 to avoid bank conflicts
+
+        int tx = blockIdx.y * tile_dim + threadIdx.x; // for loading z-direction
+        int ty = blockIdx.x * tile_dim + threadIdx.y; // for loading x-direction
```

Comment:
It looks correct; however, I find the variable names confusing. Maybe here tx could be k and ty i. Later ii = it + i.

Reply:
https://github.com/WeiqunZhang/amrex-devtests/blob/main/transpose/main.cpp This is the test I use for correctness and performance.